packages/backend/src/handlers/actions/quests/createCompletion/createCompletion.ts

```diff
@@ -0,0 +1,67 @@
+import { CreateQuestCompletionInput, CreateQuestCompletionOutput,
+  Quest_Completion_Insert_Input,
+  QuestRepetition_Enum,
+  QuestStatus_Enum, } from '../../../../lib/autogen/hasura-sdk';
+import { client } from '../../../../lib/hasuraClient';
+
+export async function createCompletion(
+  playerId: string,
+  questCompletion: CreateQuestCompletionInput,
+): Promise<CreateQuestCompletionOutput> {
+  if (!questCompletion.submission_link && !questCompletion.submission_text) {
+    throw new Error('Must provide at least a submission link or text');
+  }
+
+  const { quest_by_pk: quest } = await client.GetQuestById({ quest_id: questCompletion.quest_id });
+  if (!quest) {
+    throw new Error('Quest not found');
+  }
+
+  if (quest.status !== QuestStatus_Enum.Open) {
+    throw new Error('Quest must be open');
+  }
+
+  // Personal or unique, check if not already done by player
+  if (quest.repetition === QuestRepetition_Enum.Unique || quest.repetition === QuestRepetition_Enum.Personal) {
+    const { quest_completion: existingQuestCompletions } = await client.GetQuestCompletions({
+      player_id: playerId,
+      quest_id: questCompletion.quest_id,
+    });
+    if(existingQuestCompletions.length > 0) {
+      throw new Error('Player already completed this personal/unique quest');
```

Comment:
For these errors, we should make them user-readable:
```suggestion
      throw new Error('You have already completed this quest');
```

Reply:
ok